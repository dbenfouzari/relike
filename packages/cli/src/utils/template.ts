import {
  camelCase,
  constantCase,
  dotCase,
  paramCase,
  pascalCase,
  pathCase,
  sentenceCase,
  snakeCase,
} from 'change-case';

// String `Lives down BY the River` conversions.
export enum CaseConverter {
  camelCase = '(camelCase)',       // livesDownByTheRiver
  constantCase = '(constantCase)', // LIVES_DOWN_BY_THE_RIVER
  dotCase = '(dotCase)',           // lives.down.by.the.river
  kebabCase = '(kebabCase)',       // lives-down-by-the-river
  lowerCase = '(lowerCase)',       // livesdownbytheriver
  pascalCase = '(pascalCase)',     // LivesDownByTheRiver
  pathCase = '(pathCase)',         // lives/down/by/the/river
  sentenceCase = '(sentenceCase)', // Lives down by the river
  snakeCase = '(snakeCase)',       // lives_down_by_the_river
}

export const replaceFileName = (fileName: string, key: string, value: string) => {
  // Define my two RegExps that will allow me to replace with correct content.
  const regexWithoutCondition = new RegExp(`__${key}__`, 'gm');
  const regexWithCondition = new RegExp(`__${key}__(\\(if:\\w+\\))`, 'gm');

  // By default, value should be `kebab-case`
  value = paramCase(value);

  // Initialize my final value with initial fileName.
  let result = fileName;

  // We first have to replace the variables with case
  if (regexWithCondition.test(result)) result = result.replace(regexWithCondition, value);

  // Then we replace the final variables.
  if (regexWithoutCondition.test(result)) result = result.replace(regexWithoutCondition, value);

  result = result.replace('.txt', '');

  return result;
};

export const replace = (content: string, key: string, value: string, defaultCase = pascalCase) => {
  // Define my two RegExps that will allow me to replace with correct content.
  const regexWithoutCase = new RegExp(`__${key}__`, 'gm');
  const regexWithCase = new RegExp(`__${key}__(\\(\\w+\\))`, 'gm');

  // By default, value should be `PascalCase`
  value = defaultCase(value);

  // Initialize my final value with initial content.
  let result = content;

  // We first have to replace the variables with case
  if (regexWithCase.test(result)) result = result.replace(regexWithCase, (v, caseConverter) => {
    switch (caseConverter as CaseConverter) {
      case CaseConverter.camelCase:
        return camelCase(value);

      case CaseConverter.constantCase:
        return constantCase(value);

      case CaseConverter.dotCase:
        return dotCase(value);

      case CaseConverter.kebabCase:
        return paramCase(value);

      case CaseConverter.lowerCase:
        return value.toLowerCase().replaceAll(' ', '');

      case CaseConverter.pascalCase:
        return pascalCase(value);

      case CaseConverter.pathCase:
        return pathCase(value);

      case CaseConverter.sentenceCase:
        return sentenceCase(value);

      case CaseConverter.snakeCase:
        return snakeCase(value);

      default:
        throw new Error('Unknown case ' + caseConverter);
    }
  });

  // Then we replace the final variables.
  if (regexWithoutCase.test(result)) result = result.replace(regexWithoutCase, value);

  return result;
};
