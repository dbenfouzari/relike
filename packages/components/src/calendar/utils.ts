import DateTime from "../date-time";
import { CalendarEvent } from "./types";

/**
 * Returns all events occurring `month` month.
 *
 * It's used to get a subset of events, to be able to process fast.
 *
 * @param {DateTime} month The calendar month
 * @param {CalendarEvent[]} events The events to filter
 * @example
 * filterCalendarEvents(DateTime.now(), [])
 * @returns {CalendarEvent[]} Filtered calendar events.
 */
export const filterCalendarEvents = (month: DateTime, events: CalendarEvent[]) => {
  const firstDayOfFirstMonthWeek = month.getFirstDayOfMonthWeek();
  const lastDayOfFirstMonthWeek = month.getLastDayOfMonthWeek();

  return events.filter(
    (event) =>
      event.date.isBetween(firstDayOfFirstMonthWeek, lastDayOfFirstMonthWeek) ||
      event.endDate?.isBetween(firstDayOfFirstMonthWeek, lastDayOfFirstMonthWeek),
  );
};

/**
 * Returns all events occurring `date` day.
 *
 * You should filter events by month first to avoid processing useless data.
 *
 * @param {DateTime} date The date
 * @param {CalendarEvent[]} events The events
 * @see filterCalendarEvents
 * @example
 * getEventsThatDay(dateTime, events);
 * @returns {CalendarEvent[]} The events occurring that day
 */
export const getEventsThatDay = (date: DateTime, events: CalendarEvent[]) => {
  const start = new DateTime({
    year: date.year,
    month: date.month,
    day: date.day,
    hour: 0,
    minute: 0,
    second: 0,
  });
  const end = new DateTime({
    year: date.year,
    month: date.month,
    day: date.day,
    hour: 23,
    minute: 59,
    second: 59,
  });

  return events.filter((event) => event.date.isBetween(start, end) || event.endDate?.isBetween(start, end));
};
