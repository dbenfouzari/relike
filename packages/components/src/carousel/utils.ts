export const THRESHOLD = 0.75;

/**
 * Find the first visible carousel item index.
 *
 * @param   {HTMLElement[]} childRefs The child refs
 * @example
 * findFirstVisibleIndex(...)
 * @returns {number}                  First index
 */
export const findFirstVisibleIndex = (childRefs: HTMLElement[]) =>
  childRefs.findIndex((child) => child.getAttribute("visible"));

/**
 * Find the last visible carousel item index.
 *
 * @param   {HTMLElement[]} childRefs The child refs
 * @example
 * findLastVisibleIndex(...)
 * @returns {number}                  Last index
 */
export const findLastVisibleIndex = (childRefs: HTMLElement[]) => {
  const reversedIndex = childRefs
    // "reverse" mutates existing array, that's why we copy it via "slice"
    .slice()
    .reverse()
    .findIndex((child) => child.getAttribute("visible"));
  const count = childRefs.length - 1;
  return reversedIndex >= 0 ? count - reversedIndex : reversedIndex;
};

/**
 * Scrolls the carousel to left until it gets to defined step.
 *
 * @param {number}        scrollStep The step to go
 * @param {HTMLElement[]} itemRefs   The item refs.
 * @example
 * scrollLeftToStep(0, elms);
 */
export const scrollLeftToStep = (scrollStep: number, itemRefs: HTMLElement[]) => {
  const firstVisibleIndex = findFirstVisibleIndex(itemRefs);
  const actualScrollForIndex = firstVisibleIndex < scrollStep ? 0 : firstVisibleIndex - scrollStep;

  itemRefs[actualScrollForIndex].scrollIntoView({ behavior: "smooth" });
};

/**
 * Scrolls the carousel to right until it gets to defined step.
 *
 * @param {number}        scrollStep The step to go
 * @param {HTMLElement[]} itemRefs   The item refs.
 * @example
 * scrollRightToStep(0, elms);
 */
export const scrollRightToStep = (scrollStep: number, itemRefs: HTMLElement[]) => {
  const lastVisibleIndex = findLastVisibleIndex(itemRefs);
  const lastIndex = itemRefs.length - 1;
  const actualScrollForIndex = lastIndex - lastVisibleIndex < scrollStep ? lastIndex : lastVisibleIndex + scrollStep;
  itemRefs[actualScrollForIndex].scrollIntoView({ behavior: "smooth" });
};

/**
 * Scrolls the carousel until it gets to defined step.
 *
 * @param {number}        scrollStep The step to go
 * @param {HTMLElement[]} itemRefs   The item refs.
 * @example
 * scrollToStep(0, elms);
 */
export const scrollToStep = (scrollStep: number, itemRefs: HTMLElement[]) => {
  itemRefs[scrollStep].scrollIntoView({ behavior: "smooth" });
};

/**
 * Automatically shows or hide indicator.
 *
 * @param {HTMLElement[]} itemRefs          The item refs
 * @param {Function}      setLeftIndicator  Handler to set left indicator visibility
 * @param {Function}      setRightIndicator Handler to set right indicator visibility
 * @example
 * showHideIndicator(itemRefs, setLeftIndicator, setRightIndicator);
 */
export const showHideIndicator = (
  itemRefs: HTMLElement[],
  setLeftIndicator: (isShow: boolean) => void,
  setRightIndicator: (isShow: boolean) => void,
) => {
  const firstVisibleIndex = findFirstVisibleIndex(itemRefs);
  const lastVisibleIndex = findLastVisibleIndex(itemRefs);
  if (lastVisibleIndex === -1 && firstVisibleIndex === -1) {
    setLeftIndicator(false);
    setRightIndicator(false);
  }
  lastVisibleIndex < itemRefs.length - 1 ? setRightIndicator(true) : setRightIndicator(false);
  firstVisibleIndex > 0 ? setLeftIndicator(true) : setLeftIndicator(false);
};
