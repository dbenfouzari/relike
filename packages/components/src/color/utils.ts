/**
 * Convert r, g, and b values to HSL
 *
 * @param {number} r Red amount
 * @param {number} g Green amount
 * @param {number} b Blue amount
 * @example
 * toHsl(255, 255, 255)
 * @returns {any} HSL array
 */
export const toHsl = (r: number, g: number, b: number) => {
  // Make r, g, and b fractions of 1
  r /= 255;
  g /= 255;
  b /= 255;

  // Find greatest and smallest channel values
  const cmin = Math.min(r, g, b);
  const cmax = Math.max(r, g, b);
  const delta = cmax - cmin;

  let h, s, l;

  // Calculate hue
  // No difference
  if (delta === 0) h = 0;
  // Red is max
  else if (cmax === r) h = ((g - b) / delta) % 6;
  // Green is max
  else if (cmax === g) h = (b - r) / delta + 2;
  // Blue is max
  else h = (r - g) / delta + 4;

  h = Math.round(h * 60);

  // Make negative hues positive behind 360Â°
  if (h < 0) h += 360;

  // Calculate lightness
  l = (cmax + cmin) / 2;

  // Calculate saturation
  s = delta === 0 ? 0 : delta / (1 - Math.abs(2 * l - 1));

  // Multiply l and s by 100
  s = +(s * 100).toFixed(1);
  l = +(l * 100).toFixed(1);

  return [h, s, l];
};
