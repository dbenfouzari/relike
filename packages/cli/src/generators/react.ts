import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
