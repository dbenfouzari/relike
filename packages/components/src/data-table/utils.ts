import { Children, ReactElement } from "react";

import { DataTableColumnProps } from "./types";

/**
 * Filters DataTable children by `optional` prop.
 *
 * @param   {ReactElement[]} children The children to filter
 * @example
 * filterChildren(...)
 * @returns {ReactElement[]}          The filtered children
 */
export const filterChildren = <C extends DataTableColumnProps>(children: ReactElement<C>[]) => {
  return Children.map(children, (child) => {
    if (!child) return null;
    if (child.props.optional) return null;
    return child;
  });
};
