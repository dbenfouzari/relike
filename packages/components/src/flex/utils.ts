import { CSSProperties } from "react";

import Axis from "../axis";
import CrossAxisAlignment from "../cross-axis-alignment";
import MainAxisAlignment from "../main-axis-alignment";
import TextDirection from "../text-direction";
import VerticalDirection from "../vertical-direction";

export const getAlignItems = (crossAxisAlignment: CrossAxisAlignment): NonNullable<CSSProperties["alignItems"]> => {
  switch (crossAxisAlignment) {
    case CrossAxisAlignment.center:
    default:
      return "center";
    case CrossAxisAlignment.start:
      return "flex-start";
    case CrossAxisAlignment.end:
      return "flex-end";
    case CrossAxisAlignment.stretch:
      return "stretch";
  }
};

export const getJustifyContent = (
  mainAxisAlignment: MainAxisAlignment,
): NonNullable<CSSProperties["justifyContent"]> => {
  switch (mainAxisAlignment) {
    case MainAxisAlignment.start:
    default:
      return "flex-start";
    case MainAxisAlignment.center:
      return "center";
    case MainAxisAlignment.end:
      return "flex-end";
    case MainAxisAlignment.spaceAround:
      return "space-around";
    case MainAxisAlignment.spaceEvenly:
      return "space-evenly";
    case MainAxisAlignment.spaceBetween:
      return "space-between";
  }
};

export const getFlexDirection = (
  textDirection: TextDirection,
  verticalDirection: VerticalDirection,
  direction: Axis,
): NonNullable<CSSProperties["flexDirection"]> => {
  if (textDirection === TextDirection.ltr && direction === Axis.horizontal) return "row";
  if (textDirection === TextDirection.rtl && direction === Axis.horizontal) return "row-reverse";
  if (verticalDirection === VerticalDirection.down && direction === Axis.vertical) return "column";
  if (verticalDirection === VerticalDirection.up && direction === Axis.vertical) return "column-reverse";
  return "row";
};
