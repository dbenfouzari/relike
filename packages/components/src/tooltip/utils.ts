import { MutableRefObject } from "react";

/**
 * A function that returns nothing.
 */
export type FnRef<T> = (value: T) => void;

/**
 * Used to merge tooltip refs together.
 *
 * @param   {MutableRefObject[]} refs The refs you want to merge.
 * @example
 * mergeRefs(divRef, arrRef);
 * @returns {FnRef}                   a new ref object.
 */
export function mergeRefs<T>(...refs: Array<null | MutableRefObject<T | null> | FnRef<T>>): FnRef<T> {
  return (value) =>
    refs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref) {
        // eslint-disable-next-line no-param-reassign
        (ref as MutableRefObject<T>).current = value;
      }
    });
}
