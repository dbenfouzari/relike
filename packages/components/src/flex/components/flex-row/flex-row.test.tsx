import { Colors } from "@hastics/utils";
import { render } from "@testing-library/react";

import Container from "../../../container";
import CrossAxisAlignment from "../../../cross-axis-alignment";
import MainAxisAlignment from "../../../main-axis-alignment";
import TextDirection from "../../../text-direction";
import FlexRow, { FlexRowProps } from "./flex-row";

const defaultProps: Partial<FlexRowProps> = {
  className: "",
  crossAxisAlignment: CrossAxisAlignment.center,
  mainAxisAlignment: MainAxisAlignment.start,
  textDirection: TextDirection.ltr,
  gap: 0,
};

describe("FlexRow", () => {
  it("should render successfully", () => {
    const { container } = render(
      <FlexRow {...defaultProps}>
        <Container width={100} height={100} color={Colors.red[300]} />
        <Container width={80} height={80} color={Colors.green[300]} />
        <Container width={60} height={60} color={Colors.blue[300]} />
      </FlexRow>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
