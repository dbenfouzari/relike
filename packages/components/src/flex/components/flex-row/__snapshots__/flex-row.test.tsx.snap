// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexRow should render successfully 1`] = `
<div
  class="flex flex-0-2-1 flex-d0-0-2-2 flex__horizontal flex__cross_center flex__main_start flex__text_ltr flex__vertical_down"
>
  <div
    class="container-0-2-3 container-d0-0-2-4"
  />
  <div
    class="container-0-2-3 container-d1-0-2-5"
  />
  <div
    class="container-0-2-3 container-d2-0-2-6"
  />
</div>
`;
